### DISCUSSION TITLE
Mariner distroless 2.0 detecting false positives for telegraf

### DISCUSSION THREAD

**Discussion by jatakiajanvi12** [aquasecurity/trivy#4412] [Q&A]
### IDs

CVE-2023-28840, CVE-2023-27561, CVE-2023-28642

### Description

with Mariner base: 
Dockerfile:
FROM mcr.microsoft.com/cbl-mariner/base/core:2.0
RUN tdnf install -y telegraf
Trivy run:
[ /tmp/telegraf ]$ sudo trivy image --ignore-unfixed telegrafjanvi:latest
2023-05-10T13:50:03.144-0400    INFO    Vulnerability scanning is enabled
2023-05-10T13:50:03.144-0400    INFO    Secret scanning is enabled
2023-05-10T13:50:03.144-0400    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-05-10T13:50:03.144-0400    INFO    Please see also https://aquasecurity.github.io/trivy/v0.41/docs/secret/scanning/#recommendation for faster secret detection
2023-05-10T13:50:07.957-0400    INFO    Detected OS: cbl-mariner
2023-05-10T13:50:07.957-0400    INFO    Detecting CBL-Mariner vulnerabilities...
2023-05-10T13:50:07.979-0400    INFO    Number of language-specific files: 0
telegrafjanvi:latest (cbl-mariner 2.0.20230426)
Total: 0 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 0)

With Mariner distroless: 
Dockerfile:
FROM mcr.microsoft.com/cbl-mariner/base/core:2.0 as builder
RUN tdnf install -y telegraf
FROM mcr.microsoft.com/cbl-mariner/distroless/base:2.0  AS distroless_image
COPY --from=builder /usr/bin/telegraf /usr/bin/telegraf
Trivy run:
trivy image --ignore-unfixed --no-progress --severity HIGH,CRITICAL,MEDIUM --exit-code 1 test:new1
2023-05-10T15:05:37.587-0700    WARN    No OS package is detected. Make sure you haven't deleted any files that contain information about the installed packages.
2023-05-10T15:05:37.587-0700    WARN    e.g. files under "/lib/apk/db/", "/var/lib/dpkg/" and "/var/lib/rpm"
2023-05-10T15:05:37.587-0700    INFO    Detected OS: cbl-mariner
2023-05-10T15:05:37.587-0700    INFO    Detecting CBL-Mariner vulnerabilities...
2023-05-10T15:05:37.587-0700    INFO    Number of language-specific files: 1
2023-05-10T15:05:37.587-0700    INFO    Detecting gobinary vulnerabilities...
test:new1 (cbl-mariner 2.0.20230426)
Total: 0 (MEDIUM: 0, HIGH: 0, CRITICAL: 0)
usr/bin/telegraf (gobinary)
Total: 7 (MEDIUM: 4, HIGH: 3, CRITICAL: 0)
┌──────────────────────────────────┬────────────────┬──────────┬──────────────────────────────────────┬──────────────────┬──────────────────────────────────────────────────────────────┐
│             Library              │ Vulnerability  │ Severity │          Installed Version           │  Fixed Version   │                            Title                             │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/docker/docker         │ CVE-2023-28840 │ HIGH     │ v23.0.0+incompatible                 │ 20.10.24, 23.0.3 │ Encrypted overlay network may be unauthenticated             │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28840                   │
│                                  ├────────────────┼──────────┤                                      │                  ├──────────────────────────────────────────────────────────────┤
│                                  │ CVE-2023-28841 │ MEDIUM   │                                      │                  │ Encrypted overlay network traffic may be unencrypted         │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28841                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/docker/docker         │ CVE-2023-28842 │ MEDIUM   │ v23.0.0+incompatible                 │ 20.10.24, 23.0.3 │ Encrypted overlay network with a single endpoint is          │
│                                  │                │          │                                      │                  │ unauthenticated                                              │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28842                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/opencontainers/runc   │ CVE-2023-27561 │ HIGH     │ v1.1.4                               │ v1.1.5           │ runc: volume mount race condition (regression of             │
│                                  │                │          │                                      │                  │ CVE-2019-19921)                                              │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-27561                   │
│                                  ├────────────────┤          │                                      │                  ├──────────────────────────────────────────────────────────────┤
│                                  │ CVE-2023-28642 │          │                                      │                  │ AppArmor can be bypassed when `/proc` inside the container   │
│                                  │                │          │                                      │                  │ is symlinked with...                                         │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28642                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/opencontainers/runc   │ CVE-2023-25809 │ MEDIUM   │ v1.1.4                               │ v1.1.5           │ Rootless runc makes `/sys/fs/cgroup` writable                │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-25809                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/prometheus/prometheus │ CVE-2019-3826  │ MEDIUM   │ v1.8.2-0.20210430082741-2a4b8e12bbf2 │ v2.7.1           │ prometheus: Stored DOM cross-site scripting (XSS) attack via │
│                                  │                │          │                                      │                  │ crafted URL                                                  │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2019-3826              

Mariner doesnt use these packages while building but trivy is showing these vulnerabilities. 

### Reproduction Steps

```bash
1.Dockerfile:
FROM mcr.microsoft.com/cbl-mariner/base/core:2.0 as builder
RUN tdnf install -y telegraf
FROM mcr.microsoft.com/cbl-mariner/distroless/base:2.0  AS distroless_image
COPY --from=builder /usr/bin/telegraf /usr/bin/telegraf
2.trivy image --ignore-unfixed --no-progress --severity HIGH,CRITICAL,MEDIUM --exit-code 1 test:new1
```

### Target

Container Image

### Scanner

Vulnerability

### Target OS

Mariner 2.0 distroless

### Debug Output

```bash
trivy image --ignore-unfixed --no-progress --severity HIGH,CRITICAL,MEDIUM --exit-code 1 test:new1
2023-05-10T15:05:37.587-0700    WARN    No OS package is detected. Make sure you haven't deleted any files that contain information about the installed packages.
2023-05-10T15:05:37.587-0700    WARN    e.g. files under "/lib/apk/db/", "/var/lib/dpkg/" and "/var/lib/rpm"
2023-05-10T15:05:37.587-0700    INFO    Detected OS: cbl-mariner
2023-05-10T15:05:37.587-0700    INFO    Detecting CBL-Mariner vulnerabilities...
2023-05-10T15:05:37.587-0700    INFO    Number of language-specific files: 1
2023-05-10T15:05:37.587-0700    INFO    Detecting gobinary vulnerabilities...
test:new1 (cbl-mariner 2.0.20230426)
Total: 0 (MEDIUM: 0, HIGH: 0, CRITICAL: 0)
usr/bin/telegraf (gobinary)
Total: 7 (MEDIUM: 4, HIGH: 3, CRITICAL: 0)
┌──────────────────────────────────┬────────────────┬──────────┬──────────────────────────────────────┬──────────────────┬──────────────────────────────────────────────────────────────┐
│             Library              │ Vulnerability  │ Severity │          Installed Version           │  Fixed Version   │                            Title                             │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/docker/docker         │ CVE-2023-28840 │ HIGH     │ v23.0.0+incompatible                 │ 20.10.24, 23.0.3 │ Encrypted overlay network may be unauthenticated             │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28840                   │
│                                  ├────────────────┼──────────┤                                      │                  ├──────────────────────────────────────────────────────────────┤
│                                  │ CVE-2023-28841 │ MEDIUM   │                                      │                  │ Encrypted overlay network traffic may be unencrypted         │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28841                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/docker/docker         │ CVE-2023-28842 │ MEDIUM   │ v23.0.0+incompatible                 │ 20.10.24, 23.0.3 │ Encrypted overlay network with a single endpoint is          │
│                                  │                │          │                                      │                  │ unauthenticated                                              │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28842                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/opencontainers/runc   │ CVE-2023-27561 │ HIGH     │ v1.1.4                               │ v1.1.5           │ runc: volume mount race condition (regression of             │
│                                  │                │          │                                      │                  │ CVE-2019-19921)                                              │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-27561                   │
│                                  ├────────────────┤          │                                      │                  ├──────────────────────────────────────────────────────────────┤
│                                  │ CVE-2023-28642 │          │                                      │                  │ AppArmor can be bypassed when `/proc` inside the container   │
│                                  │                │          │                                      │                  │ is symlinked with...                                         │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-28642                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/opencontainers/runc   │ CVE-2023-25809 │ MEDIUM   │ v1.1.4                               │ v1.1.5           │ Rootless runc makes `/sys/fs/cgroup` writable                │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2023-25809                   │
├──────────────────────────────────┼────────────────┼──────────┼──────────────────────────────────────┼──────────────────┼──────────────────────────────────────────────────────────────┤
│ github.com/prometheus/prometheus │ CVE-2019-3826  │ MEDIUM   │ v1.8.2-0.20210430082741-2a4b8e12bbf2 │ v2.7.1           │ prometheus: Stored DOM cross-site scripting (XSS) attack via │
│                                  │                │          │                                      │                  │ crafted URL                                                  │
│                                  │                │          │                                      │                  │ https://avd.aquasec.com/nvd/cve-2019-3826                    │
```

### Version

```bash
trivy --version
Version: 0.28.1
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-05-16 06:07:09.707535784 +0000 UTC
  NextUpdate: 2023-05-16 12:07:09.707535384 +0000 UTC
  DownloadedAt: 2023-05-16 10:40:20.2825066 +0000 UTC
```

### Checklist

- [X] Read [the documentation regarding wrong detection](https://aquasecurity.github.io/trivy/dev/community/contribute/discussion/#false-detection)
- [ ] Ran Trivy with `-f json` that shows data sources and confirmed that the security advisory in data sources was correct

**Comment by jatakiajanvi12** [comment]
@eric-desrochers

**Reply by knqyf263** [reply]
The binary includes those dependencies.

```
$ go version -m usr/bin/telegraf | grep docker
        dep     github.com/docker/distribution  v2.8.1+incompatible
        dep     github.com/docker/docker        v23.0.0+incompatible
        dep     github.com/docker/go-connections        v0.4.0
        dep     github.com/docker/go-units      v0.5.0
```

**Reply by jatakiajanvi12** [reply]
@knqyf263 Then why only for distroless is trivy detecting these vulnerabilities? Because trivy says no vulnerabilities for the base image?

**Reply by knqyf263** [reply]
How is `telegraf` installed? If it is installed by package manager such as rpm, Trivy tries to follow security advisories from CBL-Mariner because vendors usually apply backport fixes. Otherwise, it leads to false positives. RPM database contains which files are managed by `rpm`. Trivy will skip `telegraf` if it's installed by `rpm` (`yum` or `dnf`).

One possibility is the distroless image doesn't include RPM DB, then Trivy scans `telegraf` and detect vulnerabilities according to upstream advisories such as GHSA. The base image contains RPM DB, and Trivy applies CBL-Mariner advisories, which don't provide information on the vulnerabilities,  to `telegraf` rather than upstream advisories.

**Reply by eric-desrochers** [reply]
distroless image doesn't have package manager (RPM, DNF, TDNF and all). It is as per design to limit the attack surface. We keep all the package and version in a manifest file `/var/lib/rpmmanifest/container-manifest-2` The docker image @jatakiajanvi12's team is creating is probably made from a multi-stage docker recipe that is COPYING file, binaries, ... into the distroless image and doesn't necessary update the manifest file.

@jatakiajanvi12 what your container image `/var/lib/rpmmanifest/container-manifest-2` says ?

**Reply by knqyf263** [reply]
Yes, Trivy parses the manifest file for distroless. It should include which files are installed by the package manager.

**Comment by eric-desrochers** [comment]
@jatakiajanvi12  do you have try with a later image of Mariner ? 

Picking randomly one CVE: CVE-2023-27561
```

  Library              │ Vulnerability  │ Severity │          Installed Version           │  Fixed Version 
github.com/opencontainers/runc   │ CVE-2023-27561 │ HIGH     │ v1.1.4                               │ v1.1.5 
```

In this above example, the fix is found in v1.1.5, but you seems to have v1.1.4 installed
